Prioritize recent senders / PM recipients in @-mention typeahead.

akhilmd opened the pull request:
After splitting matches into people who are subscribed and people who are not subscribed, we now split again into people with whom user has and has not been in a PM with. Then within each split, by using `message_store.recent_messages`, we prioritize recent message senders.

smarx commented:
*Automated message from Dropbox CLA bot*

@akhilmd, it looks like you've already signed the Dropbox CLA. Thanks!

akhilmd commented:
* I moved the added data structure into its own module.
* The tests fail if not for the last commit, to which I will add correct tests to match changes and squash it with the corresponding commit.

timabbott commented:
Thanks for working on this @akhilmd!  I think the new version is a lot cleaner.

One thing that I'm still worried about on looking at the first commit is that the `recent_senders` module update logic might be quadratic, since for every message, we search through all senders on that topic in a loop, looking for matches, and then sort the data structure.

I think it might be better to (1) use a dict by user ID to find the same sender within the `recent_senders` module, and (2) do the sorting by recency when we want to use the data, not when we update the data.  With those two changes, I think the update logic will be a clean `O(1)` and thus not something we need to worry about.

@akhilmd can you rework that first commit?  After you do that, I think the first commit should be mergable, though there will be a follow-up TODO item for supporting topic editing.

It looks like the remaining test failure is a flake.  

@showell can you give the second commit a review?  I feel like it's a lot more new code than I would have expected, and I'm wondering if you have an idea for a refactoring that would help.

showell commented:
I think we have some quadratic behavior here due to the indexOf() call that happens during the split/sort process.

We will probably eventually have multiple places in our code where it's useful to know that somebody is a PM conversation partner with you.  We should probably encapsulate this in a new module called pm_conversations.js, which would basically just wrap a Dict object.  The API could be something like:

- pm_conversations.set_partner(user_id)
- pm_conversations.is_partner(user_id)

We'd call `set_partner` from `message_store.process_message_for_recent_conversations`.  We'd call `is_partner` from `typeahead_helpers.split_by_pms()`.

showell commented on static/js/typeahead_helper.js:
```diff
@@ -144,12 +169,54 @@ exports.compare_by_pms = function (user_a, user_b) {
+    var subs_pm_sorted = [];
+    var non_subs_pm_sorted = [];
+    var subs_no_pm_sorted = [];
+    var non_subs_no_pm_sorted = [];
```
We use the strategy of splitting data into two arrays for sorting in a couple places in our code.  I think it's fine for totally simple cases, but it doesn't scale well to multiple splitting critieria.

This code is pretty hard to read.

I would suggest we instead use a strategy of having a compare function that triages sub-vs.non and pm-vs.non.

Here is an example:

https://github.com/zulip/zulip/blob/ee17f8fc8dcc8820dc73ab0774bb9c068fa44952/static/js/activity.js#L178

akhilmd commented:
- In recent_senders
    - Made update logic to work in constant time.
    - `get_ids_ordered_by_recency()` will sort if necessary and return a list.
- Added pm_conversations.
- Addressed readability issue,

akhilmd commented:
@timabbott 
- Rewrote recent_senders to make the processing of the message a constant time operation and shifted the responsibility of sorting to typeahead_helper.

@showell 
- Added pm_conversations to avoid calls to `indexOf` as suggested.
- Added a compare function instead of splitting.

showell commented:
I pushed the first two commits.  The third commit looks good to me, but I want @timabbott to look at it, too.

showell commented:
@akhilmd Can you address the conflicts here?

akhilmd commented:
@showell 

As discussed [here](https://chat.zulip.org/#narrow/stream/test.20suites/subject/compare.20functions.2E/near/239279), I added tests to check sorting of lists of length 2 to keep coverage at 100%.

showell commented:
I just gave this a quick skim, and the tests are really thorough!  Thanks @akhilmd.

I'll do a more detailed review tomorrow, but I wanted to get you some quick feedback.

showell commented:
Merged!  Thanks @akhilmd!

timabbott commented:
Closing, since this was merged.

showell commented:
Oops, forgot to close it.  Thanks @timabbott!
